### PR TITLE
feat(datasource/galaxy-collection): migrate from v2 to v3

### LINKED ISSUE #24981: Migrate galaxy-collection datasource from v2 to v3
### Describe the proposed change(s).

It appears that the v2 Ansible Galaxy API we've been using has stopped working and that we need to migrate to v3. However I've not been able to find a definitive reference for this year, so there is still some research to be done.

This URL for example includes the word "beta" in the examples: https://ansible.readthedocs.io/projects/galaxy-ng/en/latest/community/api_v3/

### PR DESCRIPTION
**Pull request by secustor**
## Changes
Use the v3 API instead of v2

## Context
Closes https://github.com/renovatebot/renovate/issues/24981

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

https://github.com/secustor/renovate_ansible_galaxy_test

### CODE REVIEW

File: lib/modules/datasource/galaxy-collection/index.ts [hunk 1]
```diff
@@ -55,72 +57,68 @@ export class GalaxyCollectionDatasource extends Datasource {
...
     const versionsProject = versionsUrlResponse.body;
 
-    const releases: Release[] = versionsProject.results.map((value) => {
-      const release: Release = {
-        version: value.version,
-        isDeprecated: baseProject.deprecated,
-      };
-      return release;
-    });
+    const releases = await p.map(
```
Comment (viceice) [inline lib/modules/datasource/galaxy-collection/index.ts:74]: are you sure this improves perf? there is no real parallel processing, main js loop is single threaded

**Review by viceice — COMMENTED** [review]
can we remove unused stuff from those json fixtures files?

File: lib/modules/datasource/galaxy-collection/index.ts (hunk 1 above)
Comment (secustor) [reply lib/modules/datasource/galaxy-collection/index.ts:74]: This is a left over, when the loop still contained an http request

**Comment by secustor** [comment]
> can we remove unused stuff from those json fixtures files?

Removed the biggest junks from the detailed versions already. That is the reason I have added the 1 week cache. 
Currently it is IMO a good mix between usefulness and size

**Comment by jon4hz** [comment]
If you're already migrating to galaxy v3, do you think it would be possible to enable support for custom registries aka self hosted automation hubs?

File: lib/modules/datasource/galaxy-collection/index.ts [hunk 2]
```diff
@@ -21,7 +16,7 @@ export class GalaxyCollectionDatasource extends Datasource {
 
   override readonly customRegistrySupport = false;
```
Comment (rarkins) [inline lib/modules/datasource/galaxy-collection/index.ts:17]:
```suggestion
  override readonly customRegistrySupport = true;
```

Maybe no harm in this?

**Comment by secustor** [comment]
> If you're already migrating to galaxy v3, do you think it would be possible to enable support for custom registries aka self hosted automation hubs?

Are you talking about allowing overwriting the registryUrl via packageRule or automatically detecting configured registries? 
The later would be considerable work and outside of the scope of this PR.

**Comment by jon4hz** [comment]
If I understand #14799 correctly, the manager would already detect custom registries per dependency. But I'm not very familiar with the internals of renovate, so I've no idea how difficult it would be to actually use these sources.

As for the overwrite by packageRule, that would probably be nice to have, however I'm not sure how useful that feature actually is in production unless there is a way to configure multiple urls. For example our automation hub has at least 3 different repositories (validated, rh-certified and community content), all of them have different urls and in a perfect world, renovate should be able to check them all for dependency upgrades.

**Comment by secustor** [comment]
> If I understand https://github.com/renovatebot/renovate/issues/14799 correctly, the manager would already detect custom registries per dependency. But I'm not very familiar with the internals of renovate, so I've no idea how difficult it would be to actually use these sources.

> Or the source repository in the requirements.yaml is set explicitly, in that case renovate should only check that specific repository for upgrades.

Looking at the code of `ansible-galaxy` you are right, with `customRegistrySupport = true` it should work out of the box with that refactor. 

> As for the overwrite by packageRule, that would probably be nice to have, however I'm not sure how useful that feature actually is in production unless there is a way to configure multiple urls. For example our automation hub has at least 3 different repositories (validated, rh-certified and community content), all of them have different urls and in a perfect world, renovate should be able to check them all for dependency upgrades.

If you set 3 `registryUrls` Renovate will use the `hunt` strategy for `galaxy-collection`, therefore it tries each registry until it finds a package. https://docs.renovatebot.com/configuration-options/#registryurls

**Comment by jon4hz** [comment]
Hi again,

I took another look at the API of the automation hub and unfortunately simply enabling custom registry support wont do the trick because there is a small difference in the API path when using an ah compared to galaxy.com.

For example if I have my ah repository at `https://ah.mydomain.com/api/galaxy/content/community/` the corresponding API call would be `https://ah.mydomain.com/api/galaxy/content/community/v3/plugin/ansible/content/community/collections/index/`. With this PR renovate will try to call `https://ah.mydomain.com/api/galaxy/content/community/api/v3/plugin/ansible/content/published/collections/index/` instead from what I understand.

So there are 2 problems here:

1. One "api" to much in the URL -> This could be fixed by changing the default URL to `defaultRegistryUrls = ['https://galaxy.ansible.com/api'];` and remove the "api" part in the hardcoded url paths. 
2. Renovate should replace "published" in the url path with the repository where the collection is actually stored. In my example above that repository would be "community".

**Comment by secustor** [comment]
Does this three redirect you to the proper APIs?
- https://ah.mydomain.com/api/v3/collections/{{namespace}}/{{collection}}/
- https://ah.mydomain.com/api/v3/collections/{{namespace}}/{{collection}}/versions/
- https://ah.mydomain.com/api/v3/collections/{{namespace}}/{{collection}}/versions/{{version}}/

If so I can use these instead and follow the redirects.

**Comment by jon4hz** [comment]
> Does this three redirect you to the proper APIs?

Nope, calling these endpoints returns 404. But I think that makes sense because this isn't how the automation hub works. The ah can contain multiple repositories and if my hub has e.g. 3 repositories, I have to add all 3 of them to my `ansible.cfg` for example. There is no central API to get information about all repositories at once. You always have to specific about what repo you want to use.

> Tough the documentation suggest this shouldn't be used.

I'm not sure why the galaxy API docs suggest not to use them. Until now I thought galaxy (v3) would simply be an automation hub with a single repository "published" but as it seems Redhat added some tweaks to the API. 

> Do you have API documentation for the AutomationHub?

According to https://access.redhat.com/solutions/6963967 Redhat doesn't provide public API docs. However after a quick search, I could find https://developers.redhat.com/api-catalog/api/automation-hub which seems to be the automation hubs API with the small difference that the automation hub's API actually starts with `/api/galaxy/` and not `/api/automation-hub`...

**Comment by secustor** [comment]
With the last finding I'm sure this is out of scope for this change right now. Please create a discussion and summarise the findings there.

File: lib/modules/datasource/galaxy-collection/index.ts [hunk 3]
```diff
@@ -19,9 +14,9 @@ export class GalaxyCollectionDatasource extends Datasource {
     super(GalaxyCollectionDatasource.id);
   }
 
-  override readonly customRegistrySupport = false;
+  override readonly customRegistrySupport = true;
```
Comment (viceice) [inline lib/modules/datasource/galaxy-collection/index.ts:17]: we should do this in a separate PR

File: lib/modules/datasource/galaxy-collection/index.ts [hunk 4]
```diff
@@ -35,91 +30,85 @@ export class GalaxyCollectionDatasource extends Datasource {
...
+      joinUrlParts(
+        registryUrl!,
+        'api/v3/plugin/ansible/content/published/collections/index',
+        namespace,
+        projectName,
+      ),
+    );
 
-    let baseUrlResponse: HttpResponse<BaseProjectResult>;
+    let baseUrlResponse;
```
Comment (viceice) [inline lib/modules/datasource/galaxy-collection/index.ts:42]: please add ann explicit type

File: lib/modules/datasource/galaxy-collection/index.ts [hunk 5]
```diff
@@ -35,91 +30,85 @@ export class GalaxyCollectionDatasource extends Datasource {
...
-      return null;
-    }
-
     const baseProject = baseUrlResponse.body;
 
-    const versionsUrl = `${baseUrl}versions/`;
+    const versionsUrl = ensureTrailingSlash(joinUrlParts(baseUrl, 'versions'));
 
-    let versionsUrlResponse: HttpResponse<VersionsProjectResult>;
+    let versionsUrlResponse;
```
Comment (viceice) [inline lib/modules/datasource/galaxy-collection/index.ts:53]: same

File: lib/modules/datasource/galaxy-collection/index.ts [hunk 6]
```diff
@@ -35,91 +30,85 @@ export class GalaxyCollectionDatasource extends Datasource {
...
+    const releases = versionsProject.data.map(
+      (value) => {
+        const release: Release = {
+          version: value.version,
+          releaseTimestamp: value.created_at,
+          isDeprecated: baseProject.deprecated,
+        };
+        return release;
+      },
+      { concurrency: 4 },
```
Comment (viceice) [inline lib/modules/datasource/galaxy-collection/index.ts:74]: are you sure? I don't see pmap usage here

File: lib/modules/datasource/galaxy-collection/index.ts [hunk 7]
```diff
@@ -35,91 +31,89 @@ export class GalaxyCollectionDatasource extends Datasource {
...
-
-    const releases: Release[] = versionsProject.results.map((value) => {
+    const releases = versionsProject.data.map((value) => {
       const release: Release = {
         version: value.version,
+        releaseTimestamp: value.created_at,
         isDeprecated: baseProject.deprecated,
       };
       return release;
     });
```
Comment (viceice) [inline lib/modules/datasource/galaxy-collection/index.ts:78]: why not doing this as a schema transform? Like we do on other datasources and managers?

File: lib/modules/datasource/galaxy-collection/index.ts [hunk 8]
```diff
@@ -35,91 +31,89 @@ export class GalaxyCollectionDatasource extends Datasource {
...
+    );
+    const versionDetails = response.body;
+    return {
+      version: basicRelease.version,
+      isDeprecated: basicRelease.isDeprecated,
+      downloadUrl: versionDetails.download_url,
+      newDigest: versionDetails.artifact.sha256,
+      dependencies: versionDetails.metadata.dependencies,
+      sourceUrl: versionDetails.metadata.repository,
+    };
```
Comment (viceice) [inline lib/modules/datasource/galaxy-collection/index.ts:117]: Same here, should be done as schema transform

File: lib/modules/datasource/galaxy-collection/index.ts (hunk 7 above)
Comment (secustor) [reply lib/modules/datasource/galaxy-collection/index.ts:78]: If find this more readable TBH. Tough if we want to keep this as a new convention we should add this to the dev docs

**Review by viceice — APPROVED** [review]
